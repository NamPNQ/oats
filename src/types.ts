export type TypeImplementation = string;

export type Operation = "get" | "post" | "put" | "patch" | "delete";

export const OPERATIONS: Operation[] = [
  "get",
  "post",
  "put",
  "patch",
  "delete"
];

// A PathOperation is created for each path and method combination in the
// OpenAPI doc. It represents all the information needed to print out an
// implementation of a function that makes an HTTP request to that endpoint.
export type PathOperation = {
  // e.g. "/scrapers/{id}/members"
  path: string;

  operation: Operation;

  // Corresponds to OpenAPI Operation Object summary field.
  //
  // e.g. "List all users with member privileges for a scraper target"
  summary: string;

  positionalParams: Array<{
    name: string;
    description: string;
    required: boolean;
    type: "string";
  }>;

  headerParams: Array<{
    name: string;
    description: string;
    required: boolean;
    type: "string" | "string[]" | "number" | "any";
  }>;

  queryParams: Array<{
    name: string;
    description: string;
    required: boolean;
    type: "string" | "string[]" | "number";
  }>;

  bodyParam: { name: "body"; type: TypeImplementation } | null;

  responses: {
    [code: string]: {
      description: string;

      mediaTypes: {
        [mediaType: string]: { type: TypeImplementation };
      };
    };
  };
};
