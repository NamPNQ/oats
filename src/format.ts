import { camelize } from "humps"

import { PathOperation } from "./types"
import { STATUSES } from "./statuses"

export function formatTypeField(
  readOnly: boolean,
  name: string,
  required: boolean,
  type: string
) {
  return `${readOnly ? "readonly " : ""}${name}${required ? "" : "?"}: ${type};`
}

export function isTypeNamed(type: string): boolean {
  if (type.startsWith("interface")) {
    return false
  }

  const code = type.charCodeAt(0)

  // Assume that if a type starts with an alphabetical character, then it is a
  // named type
  return (code > 64 && code < 91) || (code > 96 && code < 123)
}

export function formatTypeDeclaration(name: string, impl: string): string {
  if (!impl.startsWith("{")) {
    return `type ${name} = ${impl}`
  }

  let bracesCount = 0

  for (const c of impl) {
    if (c === "{") {
      bracesCount++
    }

    if (c === "}") {
      bracesCount--
    }

    if ((c === "|" || c === "&") && bracesCount === 0) {
      return `type ${name} = ${impl}`
    }
  }

  return `interface ${name} ${impl}`
}

export function formatPathOps(pathOps: PathOperation[]): string {
  return `
interface RequestOptions {
  signal?: AbortSignal
}

${pathOps.map(pathOp => formatPathOpTypes(pathOp)).join("\n\n")}

export class Client {
  ${pathOps.map(pathOp => formatPathOpMethod(pathOp)).join("\n\n")}

  private async request(
    method: string,
    url: string,
    params: any = {},
    options: RequestOptions = {}
  ): Promise<any> {
    const requestHeaders = this.defaultHeaders()

    for (const [k, v] of Object.entries(params.headers || {})) {
      requestHeaders.set(k, v as string)
    }

    const contentType = requestHeaders.get("Content-Type") || ""

    const body =
      params.data && contentType.includes("json")
        ? JSON.stringify(params.data)
        : params.data

    const query = params.query ? \`?\${new URLSearchParams(params.query)}\` : ""

    const response = await fetch(\`\${url}\${query}\`, {
      method,
      body,
      credentials: "same-origin",
      signal: options.signal,
      headers: requestHeaders
    })

    const { status, headers } = response
    const responseContentType = headers.get("Content-Type")

    let data

    if (responseContentType.includes("json")) {
      data = await response.json()
    } else if (responseContentType.includes("octet-stream")) {
      data = await response.blob()
    } else if (responseContentType.includes("text")) {
      data = await response.text()
    }

    return { status, headers, data }
  }

  private defaultHeaders(): Headers {
    return new Headers()
  }
}


`.trim()
}

function formatPathOpMethod(pathOp: PathOperation): string {
  return `
public ${formatName(pathOp)}(
  params: ${paramsName(pathOp)},
  options: RequestOptions = {}
): Promise<${resultName(pathOp)}> {
  return this.request(
    "${pathOp.operation.toUpperCase()}",
    ${formatURL(pathOp)},
    ${formatParamsArg(pathOp)}
    options
  ) as Promise<${resultName(pathOp)}>
}
`.trim()
}

function formatParamsArg({ bodyParam }: PathOperation): string {
  if (!bodyParam) {
    return `params,`
  }

  return `{...params, 'Content-Type': '${bodyParam.mediaType}'},`
}

function uppercase1(s: string) {
  return s.length ? `${s[0].toUpperCase()}${s.slice(1)}` : s
}

function depluralize(s: string) {
  return s.endsWith("s") ? s.slice(0, s.length - 1) : s
}

function formatName({ path, operation }: PathOperation): string {
  const nicePathName = path
    .split("/")
    .map((p, i, ps) => {
      const isGettingSingleResource =
        operation === "get" && i === ps.length - 2 && ps[i + 1].startsWith("{")

      const isAddressingSingleResource =
        operation !== "get" && i === ps.length - 1

      if (isGettingSingleResource || isAddressingSingleResource) {
        return depluralize(p)
      }

      return p
    })
    .filter(p => !p.startsWith("{"))
    .join("-")

  return camelize(`${operation}-${nicePathName}`)
}

function paramsName(pathOp: PathOperation): string {
  const pathOpName = formatName(pathOp)

  return `${uppercase1(pathOpName)}Params`
}

function resultName(pathOp: PathOperation): string {
  const pathOpName = formatName(pathOp)

  return `${uppercase1(pathOpName)}Result`
}

function resultVariantName(pathOp: PathOperation, code: string): string {
  const pathOpName = formatName(pathOp)

  return `${uppercase1(pathOpName)}${STATUSES[code]}Result`
}

function formatPathOpTypes(pathOperation: PathOperation): string {
  const paramsType = printParamsType(pathOperation)
  const responseTypes = printResultTypes(pathOperation)

  return `${paramsType}\n\n${responseTypes}`
}

function formatQueryParams({ queryParams }: PathOperation): string {
  if (!queryParams.length) {
    return ""
  }

  const fieldRequired = queryParams.some(d => d.required)

  const typeImpl = `{
  ${queryParams
    .map(d => formatTypeField(false, d.name, d.required, d.type))
    .join("\n")}
}`

  return formatTypeField(false, "query", fieldRequired, typeImpl)
}

function formatPathParams({ positionalParams }: PathOperation): string {
  if (!positionalParams.length) {
    return ""
  }

  return positionalParams
    .map(d => formatTypeField(false, d.name, d.required, d.type))
    .join("\n")
}

function formatHeadersParams({ headerParams }: PathOperation): string {
  if (!headerParams.length) {
    return ""
  }

  const fieldRequired = headerParams.some(d => d.required)

  const typeImpl = `{
  ${headerParams
    .map(d => formatTypeField(false, `"${d.name}"`, d.required, d.type))
    .join("\n")}
}`

  return formatTypeField(false, "headers", fieldRequired, typeImpl)
}

function printParamsType(pathOp: PathOperation): string {
  const dataField = pathOp.bodyParam
    ? formatTypeField(
        false,
        "data",
        pathOp.bodyParam.required,
        pathOp.bodyParam.type
      )
    : ""

  const impl = `
{
  ${formatPathParams(pathOp)}

  ${dataField}

  ${formatQueryParams(pathOp)}

  ${formatHeadersParams(pathOp)}
}
`.trim()

  return formatTypeDeclaration(paramsName(pathOp), impl)
}

function printResultTypes(pathOp: PathOperation): string {
  const variants = pathOp.responses.map(({ code, mediaTypes }) =>
    printResultType(code, mediaTypes, resultVariantName(pathOp, code))
  )

  const responseTypes = `
type ${resultName(pathOp)} =
  | ${variants.map(([name]) => name).join("\n  | ")}

${variants.map(([_, type]) => type).join("\n\n")}
`.trim()

  return responseTypes
}

function printResultType(
  code: string,
  mediaTypes: PathOperation["responses"][0]["mediaTypes"],
  name: string
): [string, string] {
  let resolvedTypeImpl = "any"

  const jsonMediaType = mediaTypes.find(d =>
    d.mediaType.includes("application/json")
  )
  const textMediaType = mediaTypes.find(d => d.mediaType.includes("text"))
  const fallbackMediaType = mediaTypes[0]

  if (jsonMediaType) {
    resolvedTypeImpl = jsonMediaType.type
  } else if (textMediaType) {
    resolvedTypeImpl = textMediaType.type
  } else if (fallbackMediaType) {
    resolvedTypeImpl = fallbackMediaType.type
  }

  const resultType = `
interface ${name} {
  status: ${code === "default" ? 500 : code};
  headers: Headers;
  data: ${resolvedTypeImpl};
}
`.trim()

  return [name, resultType]
}

function formatURL({ path, server }: PathOperation) {
  const url = server + path.replace(/{/g, "${params.")

  return url.includes("$") ? "`" + url + "`" : `"${url}"`
}
