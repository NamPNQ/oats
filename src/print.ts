import { PathOperation } from "./types"

const STATUS_CODE_NAMES = {
  200: "OK",
  201: "Created",
  202: "Accepted",
  204: "NoContent",
  400: "BadRequest",
  401: "Unauthorized",
  403: "Forbidden",
  404: "NotFound",
  413: "PayloadTooLarge",
  429: "TooManyRequests",
  500: "InternalServerError",
  503: "ServiceUnavailable",
  default: "Default"
}

function titleCase(s: string) {
  return s.length ? `${s[0].toUpperCase()}${s.slice(1).toLowerCase()}` : s
}

function depluralize(s: string) {
  return s.endsWith("s") ? s.slice(0, s.length - 1) : s
}

function niceifyName(pathName: string): string {
  return pathName
    .split("/")
    .filter(p => !p.startsWith("{"))
    .map(titleCase)
    .map((p, i, ps) => (i !== ps.length - 1 ? depluralize(p) : p))
    .join("")
}

export function printPathOperationTypes(pathOperation: PathOperation): string {
  const pathName = niceifyName(pathOperation.path)
  const typeNamePrefix = `${titleCase(
    pathOperation.operation
  )}${pathName}Params`

  const bodyField = pathOperation.bodyParam
    ? `  body${pathOperation.bodyParam.required ? "" : "?"}: ${
        pathOperation.bodyParam.type
      };\n`
    : ""

  const parametersType = `type ${typeNamePrefix} = {
${bodyField}
}`

  return `
${parametersType}
`
}

export function printPathOperation(pathOperation: PathOperation): string {
  return ""
}

export function isTypeNamed(type: string): boolean {
  const code = type.charCodeAt(0)

  // Assume that if a type starts with an alphabetical character, then it is a
  // named type
  return (code > 64 && code < 91) || (code > 96 && code < 123)
}
