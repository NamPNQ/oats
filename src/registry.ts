import { PathOperation, TypeImplementation } from "./types";

export class Registry {
  public pathOperations: PathOperation[] = [];
  public namedTypes: { [name: string]: TypeImplementation } = {};

  public registerType(name: string, impl: TypeImplementation): void {
    if (this.namedTypes[name] && this.namedTypes[name] !== impl) {
      throw new Error(
        `cannot register conflicting implementations for type "${name}"`
      );
    }

    this.namedTypes[name] = impl;
  }
}
