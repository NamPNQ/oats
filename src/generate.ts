import * as path from "path"
import { OpenAPIV3 } from "openapi-types"
import { bundle } from "swagger-parser"
import { format, resolveConfig } from "prettier"
import { get } from "lodash"

import { PathOperation, Operation, OPERATIONS } from "./types"
import { printPathOperation, printField, isTypeNamed } from "./print"

class Generator {
  private doc: OpenAPIV3.Document
  private pathOperations: PathOperation[] = []
  private namedTypes: { [name: string]: string } = {}

  constructor(doc: OpenAPIV3.Document) {
    this.doc = doc

    for (const [path, pathItemObj] of Object.entries(doc.paths)) {
      for (const op of OPERATIONS) {
        if (pathItemObj[op]) {
          this.registerPathOperation(path, op, pathItemObj[op])
        }
      }
    }
  }

  public print() {
    const namedTypesOutput = Object.entries(this.namedTypes)
      .map(([name, impl]) => `type ${name} = ${impl}`)
      .join("\n\n")

    const pathOperationsOutput = this.pathOperations
      .map(op => printPathOperation(op))
      .join("\n\n")

    const output = `${namedTypesOutput}\n\n${pathOperationsOutput}`

    return output
  }

  private registerPathOperation(
    path: string,
    operation: Operation,
    operationObj: OpenAPIV3.OperationObject
  ) {
    const parameters: OpenAPIV3.ParameterObject[] = (
      operationObj.parameters || []
    ).map(p => this.followReference(p))

    const pathOperation: PathOperation = {
      path,
      operation,
      summary: operationObj.summary,
      positionalParams: this.collectParameters(parameters, "path"),
      headerParams: this.collectParameters(parameters, "header"),
      queryParams: this.collectParameters(parameters, "query"), // TODO: Support "string[]" parameters
      bodyParam: this.collectBodyParam(operationObj),
      responses: this.collectResponses(operationObj.responses)
    }

    this.pathOperations.push(pathOperation)
  }

  private collectParameters(
    allParameters: OpenAPIV3.ParameterObject[],
    parameterLocation: "query" | "path" | "header"
  ) {
    return allParameters
      .filter(p => p.in === parameterLocation)
      .map(p => ({
        name: p.name,
        description: p.description,
        required: p.required,
        type: this.expectSimpleType(p)
      }))
  }

  private collectBodyParam(
    operationObj: OpenAPIV3.OperationObject
  ): PathOperation["bodyParam"] {
    if (!operationObj.requestBody) {
      return null
    }

    const requestBody: OpenAPIV3.RequestBodyObject = this.followReference(
      operationObj.requestBody
    )

    const jsonContent = get(requestBody, ["content", "application/json"])
    const { description, required = false } = requestBody

    if (!jsonContent) {
      return {
        description,
        required,
        type: "any"
      }
    }

    return {
      description,
      required,
      type: this.getType(jsonContent.schema)
    }
  }

  private collectResponses(
    responses?: OpenAPIV3.ResponsesObject
  ): PathOperation["responses"] {
    if (!responses) {
      return null
    }

    return Object.entries(responses).reduce(
      (acc, [responseCode, refOrResponse]) => {
        const responseObj: OpenAPIV3.ResponseObject = this.followReference(
          refOrResponse
        )

        if (!responseObj.content) {
          // If a response code is defined but
          return {
            ...acc,
            [responseCode]: {
              description: responseObj.description,
              mediaTypes: null
            }
          }
        }

        return {
          ...acc,
          [responseCode]: {
            description: responseObj.description,
            mediaTypes: Object.entries(responseObj.content).reduce(
              (acc, [mediaType, mediaTypeObj]) => ({
                ...acc,
                [mediaType]: {
                  type: this.getType(mediaTypeObj.schema)
                }
              }),
              {}
            )
          }
        }
      },
      {}
    )
  }

  private followReference(obj) {
    if ("$ref" in obj) {
      return get(this.doc, obj.$ref.split("/").slice(1))
    }

    return obj
  }

  private expectSimpleType(obj): "string" | "number" | "any" {
    switch (get(obj, "schema.type", null)) {
      case "string":
        return "string"
      case "integer":
      case "number":
        return "number"
      default:
        return "any"
    }
  }

  private getRefName(refString: string): string {
    const refParts = refString.split("/")
    const name = refParts[refParts.length - 1]

    return name
  }

  private getType(
    obj: OpenAPIV3.ReferenceObject | OpenAPIV3.SchemaObject
  ): string {
    if ("$ref" in obj) {
      const typeName = this.getRefName(obj.$ref)
      const schema = this.followReference(obj)
      const typeImpl = this.getTypeFromSchemaObj(schema)

      this.registerType(typeName, typeImpl)

      return typeName
    }

    return this.getTypeFromSchemaObj(obj)
  }

  private getTypeFromSchemaObj(obj: OpenAPIV3.SchemaObject): string {
    switch (obj.type) {
      case "number":
      case "integer":
        return "number"
      case "boolean":
        return "boolean"
      case "string":
        return obj.enum ? obj.enum.map(v => `"${v}"`).join(" | ") : "string"
      case "null":
        return "null"
      case "array":
        return this.getTypeFromArraySchemaObj(obj)
      case "object":
      default:
        return this.getTypeFromObjectSchemaObj(obj)
    }
  }

  private getTypeFromArraySchemaObj(obj: OpenAPIV3.ArraySchemaObject): string {
    const type = this.getType(obj.items)

    return isTypeNamed(type) ? `${type}[]` : `Array<${type}>`
  }

  private getTypeFromObjectSchemaObj(
    obj: OpenAPIV3.NonArraySchemaObject
  ): string {
    if (obj.allOf) {
      return this.getTypeFromAllOf(obj)
    }

    if (obj.oneOf) {
      return this.getTypeFromOneOf(obj)
    }

    if (obj.anyOf) {
      return this.getTypeFromOneOf(obj)
    }

    if (obj.properties) {
      return this.getTypeFromPropertiesSchemaObj(obj)
    }

    return "any"
  }

  private getTypeFromPropertiesSchemaObj(
    obj: OpenAPIV3.NonArraySchemaObject
  ): string {
    const fields = Object.entries(obj.properties).map(([name, value]) => {
      const readOnly = (value as any).readOnly
      const required = (obj.required || []).includes(name)
      const description = (value as any).description

      return printField(
        readOnly,
        name,
        required,
        this.getType(value),
        description
      )
    })

    return `{\n  ${fields.join("\n  ")}\n}`
  }

  private getTypeFromAllOf(obj: OpenAPIV3.SchemaObject): string {
    return obj.allOf.map(childObj => this.getType(childObj)).join(" & ")
  }

  private getTypeFromOneOf(obj: OpenAPIV3.SchemaObject): string {
    return obj.oneOf
      .map(childObj => {
        const type = this.getType(childObj)

        if (obj.discriminator) {
          const mappingKey = Object.keys(obj.discriminator).find(
            key => obj.discriminator.mapping[key] === (childObj as any).$ref
          )

          const propertyValue = mappingKey ? `"${mappingKey}"` : "string"

          return `(${type} & {${obj.discriminator.propertyName}: ${propertyValue}})`
        }

        return type
      })
      .join(" | ")
  }

  private registerType(name: string, impl: string): void {
    if (this.namedTypes[name] && this.namedTypes[name] !== impl) {
      throw new Error(
        `cannot register conflicting implementations for type "${name}"`
      )
    }

    this.namedTypes[name] = impl
  }
}

export async function generate(
  docOrPathToDoc: string | OpenAPIV3.Document
): Promise<string> {
  const doc = (await bundle(docOrPathToDoc)) as OpenAPIV3.Document

  const prettierConfig = await resolveConfig(path.resolve(__dirname, "../.."))

  const output = format(new Generator(doc).print(), {
    ...prettierConfig,
    parser: "typescript"
  })

  return output
}
