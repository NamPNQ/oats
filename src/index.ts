import { generate } from "./generate";

generate(process.argv[2]).then(result => console.log(result));
