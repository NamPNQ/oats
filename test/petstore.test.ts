import { generate } from "../src"

test.only("petstore", async () => {
  const expected = ""

  const actual = await generate("./fixtures/petstore.yml")

  expect(actual.trim()).toEqual(expected.trim())
})
