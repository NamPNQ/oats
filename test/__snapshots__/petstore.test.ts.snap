// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`petstore 1`] = `
"// This file is generated by [oats][0] and should not be edited by hand.
//
// [0]: https://github.com/chnn/oats

export type Pets = Pet[];

export type Pet = Dog | Cat;

export type Dog = BasePet & {
  weight: number;
};

export interface BasePet {
  id: number;
  name: string;
  category?: number;
  tag?: string;
  sex?: \\"male\\" | \\"female\\";
}

export type Cat = BasePet & {
  fluffy: boolean;
};

export interface Error {
  code: number;
  message: string;
}

interface RequestOptions {
  signal?: AbortSignal;
}

const request = async (
  method: string,
  url: string,
  params: any = {},
  options: RequestOptions = {}
): Promise<any> => {
  const requestHeaders = new Headers(params.headers);
  const contentType = requestHeaders.get(\\"Content-Type\\") || \\"\\";

  const body =
    params.data && contentType.includes(\\"json\\")
      ? JSON.stringify(params.data)
      : params.data;

  const query = params.query ? \`?\${new URLSearchParams(params.query)}\` : \\"\\";

  const response = await fetch(\`\${url}\${query}\`, {
    method,
    body,
    credentials: \\"same-origin\\",
    signal: options.signal,
    headers: requestHeaders
  });

  const { status, headers } = response;
  const responseContentType = headers.get(\\"Content-Type\\");

  let data;

  if (responseContentType.includes(\\"json\\")) {
    data = await response.json();
  } else if (responseContentType.includes(\\"octet-stream\\")) {
    data = await response.blob();
  } else if (responseContentType.includes(\\"text\\")) {
    data = await response.text();
  }

  return { status, headers, data };
};

export interface GetPetsParams {
  query?: {
    limit?: number;
  };
}

type GetPetsResult = GetPetsOKResult | GetPetsDefaultResult;

interface GetPetsOKResult {
  status: 200;
  headers: Headers;
  data: Pets;
}

interface GetPetsDefaultResult {
  status: 500;
  headers: Headers;
  data: Error;
}

export const getPets = (
  params: GetPetsParams,
  options: RequestOptions = {}
): Promise<GetPetsResult> =>
  request(
    \\"GET\\",
    \\"http://petstore.swagger.io/v1/pets\\",
    params,
    options
  ) as Promise<GetPetsResult>;

export interface PostPetParams {
  data: Pet;
}

type PostPetResult = PostPetOKResult | PostPetDefaultResult;

interface PostPetOKResult {
  status: 200;
  headers: Headers;
  data: {
    pet: Pet;
  };
}

interface PostPetDefaultResult {
  status: 500;
  headers: Headers;
  data: Error;
}

export const postPet = (
  params: PostPetParams,
  options: RequestOptions = {}
): Promise<PostPetResult> =>
  request(
    \\"POST\\",
    \\"http://petstore.swagger.io/v1/pets\\",
    { ...params, headers: { \\"Content-Type\\": \\"application/json\\" } },
    options
  ) as Promise<PostPetResult>;

export interface GetPetParams {
  petID: string;
}

type GetPetResult = GetPetOKResult | GetPetDefaultResult;

interface GetPetOKResult {
  status: 200;
  headers: Headers;
  data: Pets;
}

interface GetPetDefaultResult {
  status: 500;
  headers: Headers;
  data: Error;
}

export const getPet = (
  params: GetPetParams,
  options: RequestOptions = {}
): Promise<GetPetResult> =>
  request(
    \\"GET\\",
    \`http://petstore.swagger.io/v1/pets/\${params.petID}\`,
    params,
    options
  ) as Promise<GetPetResult>;
"
`;
